stdpopsim/cache.py

```diff
@@ -43,3 +49,128 @@ def get_cache_dir():
 
 
 set_cache_dir()
+
+
+@attr.s(kw_only=True)
+class CachedData:
+    """
+    Downloadable data that will be cached locally.
+
+    The downloadable should be a single file. The local cache may
+    correspond to this same file, or to its extracted contents. In the latter
+    case, the downloaded file will be removed after archive extraction.
+    The downloaded file is compared against the expected SHA256 checksum,
+    and if correct, the checksum is then also stored locally.
+
+    :ivar namespace: The namespace under which the cache will be stored.
+        This will be converted into a folder, by constructing folders in the
+        cache corresponding to each component of the namespace.
+        E.g. if we're on a unix system with cache under ``/path/to/cache``, and
+        ``namespace=pathlib.PurePath("foo", "bar")``, the cached data will
+        live under ``/path/to/cache/foo/bar``.
+    :vartype namespace: :class:`pathlib.PurePath`
+    :ivar str url: The URL of the data to be cached.
+    :ivar str sha256: The SHA256 checksum of the downloaded file.
+    :ivar bool extract: True if the downloaded file is a tarball that should be
+        extracted into the cached namespace, False otherwise.
+    """
+    namespace = attr.ib(type=pathlib.PurePath)
```

Comment:
I wonder if we're abusing PurePath a little bit here, since the namespace will never correspond to either a relative or absolute path in the filesystem. It's sort of a "path segment".

What are we gaining here from making the type formally a PurePath vs a string?

Reply:
I think this will be a relative path in the filesystem though. I wanted to recreate the use case of `namespace="genetic_maps/HomSap/HapMapII_GRCh37"`, which will be a path relative to the cache dir. But maybe we can do `pathlib.Path("something") / "foo/bar/baz"` on Windows anyway? Or maybe its better to make the ``namespace`` param a tuple?

Reply:
Indeed we can use forward slashes with pathlib on windows without problem. And pathlib is so awesome we can actually check that without needing a windows system.
```python
>>> from pathlib import PureWindowsPath
>>> PureWindowsPath("C:\\")
PureWindowsPath('C:/')
>>> PureWindowsPath("C:\\") / "foo/bar"
PureWindowsPath('C:/foo/bar')
>>> str(PureWindowsPath("C:\\") / "foo/bar")
'C:\\foo\\bar'
```